internal/allocator/allocator.go

```diff
@@ -194,113 +187,118 @@ func (a *Allocator) Unassign(svc string) bool {
 		}
-		delete(a.portsInUse[al.ip.String()], port)
-	}
-	delete(a.servicesOnIP[al.ip.String()], svc)
-	if len(a.portsInUse[al.ip.String()]) == 0 {
-		delete(a.portsInUse, al.ip.String())
-		delete(a.sharingKeyForIP, al.ip.String())
-	}
-	a.poolIPsInUse[al.pool][al.ip.String()]--
-	if a.poolIPsInUse[al.pool][al.ip.String()] == 0 {
-		// Explicitly delete unused IPs from the pool, so that len()
-		// is an accurate count of IPs in use.
-		delete(a.poolIPsInUse[al.pool], al.ip.String())
 	}
 	stats.poolActive.WithLabelValues(al.pool).Set(float64(len(a.poolIPsInUse[al.pool])))
 	return true
 }
 
-func cidrIsIPv6(cidr *net.IPNet) bool {
-	return cidr.IP.To4() == nil
-}
-func ipIsIPv6(ip net.IP) bool {
-	return ip.To4() == nil
-}
-
 // AllocateFromPool assigns an available IP from pool to service.
-func (a *Allocator) AllocateFromPool(svc string, isIPv6 bool, poolName string, ports []Port, sharingKey, backendKey string) (net.IP, error) {
+func (a *Allocator) AllocateFromPool(svc string, serviceIPFamily ipfamily.IPFamily, poolName string, ports []Port, sharingKey, backendKey string) ([]net.IP, error) {
 	if alloc := a.allocated[svc]; alloc != nil {
 		// Handle the case where the svc has already been assigned an IP but from the wrong family.
-		// This "should-not-happen" since the "ipFamily" is an immutable field in services.
-		if isIPv6 != ipIsIPv6(alloc.ip) {
-			return nil, fmt.Errorf("IP for wrong family assigned %s", alloc.ip.String())
+		// This "should-not-happen" since the "serviceIPFamily" is an immutable field in services.
+		if allocipsFamily, err := ipfamily.IPFamilyForAddressesIPs(alloc.ips); err != nil || allocipsFamily != serviceIPFamily {
+			return nil, fmt.Errorf("IP for wrong family assigned alloc %s service family %s", allocipsFamily, serviceIPFamily)
 		}
-		if err := a.Assign(svc, alloc.ip, ports, sharingKey, backendKey); err != nil {
+		if err := a.Assign(svc, alloc.ips, ports, sharingKey, backendKey); err != nil {
 			return nil, err
 		}
-		return alloc.ip, nil
+		return alloc.ips, nil
 	}
 
 	pool := a.pools[poolName]
 	if pool == nil {
 		return nil, fmt.Errorf("unknown pool %q", poolName)
 	}
+	sk := &key{
+		sharing: sharingKey,
+		backend: backendKey,
+	}
+	ips := []net.IP{}
+	v4ToAdd := true
+	v6ToAdd := true
 
+	switch serviceIPFamily {
+	case ipfamily.IPv4:
+		v6ToAdd = false
+	case ipfamily.IPv6:
+		v4ToAdd = false
+	}
 	for _, cidr := range pool.CIDR {
-		if cidrIsIPv6(cidr) != isIPv6 {
+		if (ipfamily.IPFamilyForCIDR(cidr) == ipfamily.IPv4 && !v4ToAdd) ||
+			(ipfamily.IPFamilyForCIDR(cidr) == ipfamily.IPv6 && !v6ToAdd) {
 			// Not the right ip-family
 			continue
 		}
-		c := ipaddr.NewCursor([]ipaddr.Prefix{*ipaddr.NewPrefix(cidr)})
-		for pos := c.First(); pos != nil; pos = c.Next() {
-			ip := pos.IP
-			if pool.AvoidBuggyIPs && ipConfusesBuggyFirmwares(ip) {
-				continue
+		ip := a.getIPFromCIDR(cidr, pool.AvoidBuggyIPs, svc, ports, sk)
+		if ip != nil {
+			ips = append(ips, ip)
```

Comment:
by doing this, we'll always append ips regardless of the family.
You need to skip if the ip is already added (or not requested):

```
                        if ipfamily.IPFamilyForCIDR(cidr) == ipfamily.IPv4 && !v4ToAdd {
				continue
			}
                        if ipfamily.IPFamilyForCIDR(cidr) == ipfamily.IPv6 && !v6ToAdd {
				continue
			}
                        ips = append(ips, ip)
```

As a second thought, maybe using a map[ipFamily]bool instead of two flags will make things more clean?

Reply:
Sorry, did not notice https://github.com/metallb/metallb/pull/1065/files#diff-e11c83ce04539e8dc3dfc6cb1227f7057e4be177233b6d6765c4b3b1691e7411R247

The logic is good, I think using a map will make it slimmer.